pkg/visitors/createfv.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2018 the LinuxBoot Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package visitors
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"strconv"
+	"unsafe"
+
+	"github.com/linuxboot/fiano/pkg/guid"
+	"github.com/linuxboot/fiano/pkg/uefi"
+)
+
+// CreateFV creates a firmware volume at given offset
+type CreateFV struct {
+	AbsOffset uint64
+	Size      uint64
+	Name      guid.GUID
+
+	offset uint64
+	found  bool
+}
+
+// Run wraps Visit and performs some setup and teardown tasks.
+func (v *CreateFV) Run(f uefi.Firmware) error {
+	err := f.Apply(v)
+	if err != nil {
+		return err
+	}
+	if !v.found {
+		return fmt.Errorf("Cannot create FV at %#x (+%#x) no BIOS region found", v.AbsOffset, v.Size)
+	}
+	return nil
+}
+
+// Visit applies the Remove visitor to any Firmware type.
+func (v *CreateFV) Visit(f uefi.Firmware) error {
+	if v.found {
+		return nil
+	}
+	var offset, end uint64
+	switch f := f.(type) {
+	case *uefi.BIOSRegion:
+		if f.FRegion != nil {
+			offset = uint64(f.FRegion.BaseOffset())
+			end = uint64(f.FRegion.EndOffset())
+		} else {
+			end = f.Length
+		}
+		if v.AbsOffset < offset {
+			return fmt.Errorf("Cannot create FV at %#x, BIOS region starts at %#x", v.AbsOffset, offset)
+		}
+		if v.AbsOffset+v.Size >= end {
+			return fmt.Errorf("Cannot create FV ending at %#x (%#x + %#x), BIOS region ends at %#x", v.AbsOffset+v.Size, v.AbsOffset, v.Size, end)
+		}
+
+		// Manually visit children, we want the index
+		idx := -1
+		var bp *uefi.BIOSPadding
+	L:
+		for i, e := range f.Elements {
+			switch f := e.Value.(type) {
+			case *uefi.BIOSPadding:
+				bp = f
+				bpOffset := offset + f.Offset
+				bpEnd := bpOffset + uint64(len(f.Buf()))
+				if v.AbsOffset >= bpOffset && v.AbsOffset+v.Size <= bpEnd {
```

Comment:
This won't work if for whatever reason we have two biospads that are contiguous (Not sure why that'd be the case though, but it might happen if we delete FVs in the future.)

Reply:
I'm not happy about the way I did biospads in the past, maybe it's time to revisit them and remove biospads. Alternatively we should do something that joins contiguous biospad regions so that this won't happen.

Reply:
I see ... Shall I leave a TODO here ?
I'll put a comment just warning about that

Reply:
Yeah leave a TODO here.